### PR TITLE
[WIP] GitHub code: tar processing

### LINKED ISSUE #3030: [Caloris] [Github Code] Tar download and processing
Periodic tar download and processing

Apply limits: 1GB / repo, 1mb / file
Pretty strict blacklist, maybe even whitelist

See https://github.com/dust-tt/dust/issues/2324

### PR DESCRIPTION
**Pull request by spolu**
Fixes https://github.com/dust-tt/dust/issues/3030

Function to:
- downloads
- filters
- preprocess files to build all necessary information for aggregation

The cli will let me check what our repo looks like

### CODE REVIEW

**Comment by spolu** [comment]
Left to do is to create directories objects so that it's super easy to use from the activity that will call into it.

File: connectors/src/connectors/github/lib/github_api.ts [hunk 1]
```diff
@@ -526,3 +535,244 @@ async function getOctokit(installationId: string): Promise<Octokit> {
...
+    repo: repoName,
+  });
+  const defaultBranch = data.default_branch;
+
+  let { data: tarballStream } = await octokit.request(
+    "GET /repos/{owner}/{repo}/tarball/{ref}",
+    {
+      owner: login,
+      repo: repoName,
+      ref: defaultBranch,
```
Comment (flvndvd) [inline connectors/src/connectors/github/lib/github_api.ts:634]: Reading the code one can provide `"parseSuccessResponseBody": false` to have a stream. See https://github.com/octokit/request.js?tab=readme-ov-file#request.

File: connectors/src/connectors/github/lib/github_api.ts [hunk 2]
```diff
@@ -526,3 +535,244 @@ async function getOctokit(installationId: string): Promise<Octokit> {
...
+  if (tarballStream instanceof ArrayBuffer) {
+    // Wrap ArrayBuffer with a stream
+    const stream = new Readable();
+    stream.push(Buffer.from(tarballStream));
+    stream.push(null); // Signal the end of the stream
+    tarballStream = stream;
+  }
+
+  // Save the tarball to the temp directory.
+  await asyncPipeline(tarballStream as Readable, createWriteStream(tarPath));
```
Comment (flvndvd) [inline connectors/src/connectors/github/lib/github_api.ts:652]: `nit` pipeline are useful when there is an operation in between read and write. Here a simple `pipe` could do the trick:

Something like:
```
const fileStream = createWriteStream(tarPath);
request(<gh_link>).pipe(fileStream);
```

File: connectors/src/connectors/github/lib/github_api.ts [hunk 3]
```diff
@@ -526,3 +535,244 @@ async function getOctokit(installationId: string): Promise<Octokit> {
...
+    internalId: string;
+    parentInternalId: string | null;
+    parents: string[];
+  }[] = [];
+
+  // Iterate over the files in the temp directory.
+  for await (const file of getFiles(tempDir)) {
+    console.log(file);
+    // get file extension
+    const ext = file.split(".").pop();
```
Comment (flvndvd) [inline connectors/src/connectors/github/lib/github_api.ts:687]: `nit`

```suggestion
    const ext = path.extname(file);
```

File: connectors/src/connectors/github/lib/github_api.ts [hunk 4]
```diff
@@ -526,3 +535,244 @@ async function getOctokit(installationId: string): Promise<Octokit> {
...
+    // get file size
+    const { size } = await fs.stat(file);
+
+    const isWithelisted =
+      EXTENSION_WHITELIST.includes(`.${ext}`) ||
+      FILENAME_WHITELIST.includes(file);
+
+    const isUnderLimite = size < 1024 * 1024;
```
Comment (flvndvd) [inline connectors/src/connectors/github/lib/github_api.ts:695]:
```suggestion
    const isUnderLimit = size < 1024 * 1024;
```

File: connectors/src/connectors/github/lib/github_api.ts [hunk 5]
```diff
@@ -526,3 +535,244 @@ async function getOctokit(installationId: string): Promise<Octokit> {
...
+        );
+      }
+
+      const documentId = `github-code-${repoId}-file-${blake3(
+        `github-code-${repoId}-file-${path.join("/")}/${file}`
+      )
+        .toString("hex")
+        .substring(0, 16)}`;
+
+      const fileName = file.split("/").pop() || "";
```
Comment (flvndvd) [inline connectors/src/connectors/github/lib/github_api.ts:720]: You could abstract all of this by using the native `path` library.

File: connectors/src/connectors/github/lib/github_api.ts [hunk 6]
```diff
@@ -526,3 +535,244 @@ async function getOctokit(installationId: string): Promise<Octokit> {
...
+    } else {
+      yield res;
+    }
+  }
+}
+
+// This function returns file and directories object with parent, internalIds, and sourceUrl
+// information. The root of the directory is considered the null parent (and will have to be
+// stitched by the activity).
+export async function processRepository(
```
Comment (flvndvd) [inline connectors/src/connectors/github/lib/github_api.ts:615]: `nit` maybe split this function a bit.

File: connectors/src/admin/cli.ts [hunk 7]
```diff
@@ -99,6 +103,63 @@ const connectors = async (command: string, args: parseArgs.ParsedArgs) => {
...
+      });
+
+      console.log(
+        `Found ${files.length} files in ${directories.length} directories`
+      );
+      console.log(
+        `Files total size: ${files.reduce((acc, f) => acc + f.sizeBytes, 0)}`
+      );
+
+      await cleanUpProcessRepository(tempDir);
```
Comment (flvndvd) [inline connectors/src/admin/cli.ts:158]: Should we wrap the code above with a `try/catch/finally` so we always remove it afterward?

**Review by flvndvd — COMMENTED** [review]
LGTM, left a couple of comments.

File: connectors/src/connectors/github/lib/github_api.ts (hunk 1 above)
Comment (flvndvd) [reply connectors/src/connectors/github/lib/github_api.ts:634]: Example: https://github.com/octokit/octokit.js/issues/2369

File: connectors/src/admin/cli.ts (hunk 7 above)
Comment (spolu) [reply connectors/src/admin/cli.ts:158]: This is just for testing and the only thing that can really fail is processRepository, but if it fails we don't know the tempDir :/

I'll add a clean-up in a try/catch/finally in processRepository

**Review by flvndvd — APPROVED** [review]

**Pull request merged**